### PR TITLE
Fix TTL of SOA record for negative answers

### PR DESCRIPTION
**Pull request by gthess**
This affects localzone and authzone data. The TTL is now the minimum of the SOA TTL and the SOA.MINIMUM.

### CODE REVIEW

File: testdata/auth_zonefile_down.rpl [hunk 1]
```diff
@@ -1,6 +1,12 @@
 ; config options
 server:
 	target-fetch-policy: "0 0 0 0 0"
+	; Options for singed zone. The zone is partially copied from val_negcache_nxdomain.rpl
```
Comment (wcawijngaards) [inline testdata/auth_zonefile_down.rpl]:
```suggestion
	; Options for signed zone. The zone is partially copied from val_negcache_nxdomain.rpl
```

File: services/localzone.c [hunk 2]
```diff
@@ -463,6 +463,48 @@ lz_find_create_node(struct local_zone* z, uint8_t* nm, size_t nmlen,
...
+	rrset_negative->rk.rrset_class = soa_rrset->rk.rrset_class;
+	if(!rrset_insert_rr(z->region, pd, rdata, rdata_len, ttl, rrstr))
+		return 0;
+	/* last 4 bytes are minimum ttl in network format */
+	if(pd->count == 0 || pd->rr_len[0] < 2+4) {
+		return 0;
+	}
+	minimum = (time_t)sldns_read_uint32(pd->rr_data[0]+(pd->rr_len[0]-4));
+	pd->ttl = ttl<minimum?ttl:minimum;
+	pd->rr_ttl[0] = pd->ttl;
```
Comment (wcawijngaards) [inline services/localzone.c]: You are missing out on the rrsig ttl's here, that are fixed in the authzone code.

**Review by wcawijngaards — APPROVED** [review]
Nice stuff, approve of it!

File: testdata/auth_zonefile_down.rpl (hunk 1 above)
Comment (gthess) [reply testdata/auth_zonefile_down.rpl]: Fixed

File: services/localzone.c (hunk 2 above)
Comment (gthess) [reply services/localzone.c]: I am not sure if rrsigs are put under the relevant rrsets for local data. From what I see in the code and what I tried I can't get an rrsig answer from local data, unless I ask for the rrsig type.

**Comment by gthess** [comment]
Thanks Wouter! One comment is still unresolved but I don't think it can be resolved unless I am missing something. After that I am ready to merge.

File: services/localzone.c (hunk 2 above)
Comment (wcawijngaards) [reply services/localzone.c]: Yes you must be right, that solves it.

**Pull request merged**